ui: Statements tracing

koorosh opened the pull request:
Depends on: #45576, #45574, #45492, #45491, #45533

Resolves: #45245, #45249, #45247, #45250

This is the final PR in chain of PRs related to Statements tracing functionality.
It is based on all intermediate PRs and in addition contains minor enhancements
 which were forgotten in previous PRs as well.

- Show alert message after user requests statement tracing report;
- Add sorting for Diagnostics tables;
- Add correct URLs on documentation for statement diagnostics;
- View button is removed because we do not have API for this;
- Add Download file behaviour on Statement diagnostics history page;

koorosh committed:
ui: (fix) Apply styles for navigation panes

Before, Text component styles were adjusted to style
links on side navigation panel. It was wrong because
Text component should not care about particular styling
of navigation pane.
Now, particular styles which relate to navigation panel
only is moved to SideNavigation component.
It is changed now because current implementation breaks
styles for Diagnostics panel.

Release note: None

koorosh committed:
ui: Link component

Link component is a wrapper on top of <a> element
is created to maintain consistent styles for links.

Release note: None

koorosh committed:
ui: Add Diagnostics view for Statements

- Adjust SummaryCard to provide external styles
- Add DiagnosticsView component which renders
view when no diagnostics available.
- Current implementation includes only view without
any logic connected to redux store.

Release note: None

koorosh committed:
ui: Add Diagnostics view for Statements

- Adjust SummaryCard to provide external styles
- Add DiagnosticsView component which renders
view when no diagnostics available.
- Current implementation includes only view without
any logic connected to redux store.

Release note: None

koorosh committed:
ui: Add Diagnostics view for Statements

- Adjust SummaryCard to provide external styles
- Add DiagnosticsView component which renders
view when no diagnostics available.
- Current implementation includes only view without
any logic connected to redux store.

Release note: None

koorosh committed:
ui: Add sagas to handle statement tracing actions

Statements sagas are implemented to be able request
statement diagnostics and retrieve when its ready.

This commit defines Actions for requesting diagnostics
and sagas which listens for those actions.

Also selectors are added to find statement by
statement id (query) which contains diagnostic data.

Release note: None

koorosh committed:
ui: Diagnostics tab with tracing records

- Add table to represent diagnostics data of statement
- Connect Statements view with statement selector to properly
display tab header (whether diagnostics data is present or not)

Release note (admin ui): Add Diagnostics tab for statement details page

koorosh committed:
ui: Diagnostics tab with tracing records

- Add table to represent diagnostics data of statement
- Connect Statements view with statement selector to properly
display tab header (whether diagnostics data is present or not)

Release note (admin ui): Add Diagnostics tab for statement details page

koorosh committed:
ui: Diagnostics tab with tracing records

- Add table to represent diagnostics data of statement
- Connect Statements view with statement selector to properly
display tab header (whether diagnostics data is present or not)

Release note (admin ui): Add Diagnostics tab for statement details page

koorosh committed:
ui: Diagnostics tab with tracing records

- Add table to represent diagnostics data of statement
- Connect Statements view with statement selector to properly
display tab header (whether diagnostics data is present or not)

Release note (admin ui): Add Diagnostics tab for statement details page

koorosh committed:
ui: Add tooltips for diagnostics statuses

- All code related to statements diagnostics are moved to
statements/diagnostics dir;
- Status badges are refactored to external component which is
responsible for displaying proper text of status and text for
tooltips.
- Tooltips have to be shown on some views and disabled on other.
To implement this, `enableTooltip` property is added. Particulary
it is needed to hide tooltip for "READY" status in Diagnostics tab
and show it on Statements List view.

Release note (admin ui change): Show tooltips for statement diagnostic statuses

koorosh committed:
ui: Add tooltips for diagnostics statuses

- All code related to statements diagnostics are moved to
statements/diagnostics dir;
- Status badges are refactored to external component which is
responsible for displaying proper text of status and text for
tooltips.
- Tooltips have to be shown on some views and disabled on other.
To implement this, `enableTooltip` property is added. Particulary
it is needed to hide tooltip for "READY" status in Diagnostics tab
and show it on Statements List view.

Release note (admin ui): Show tooltips for statement diagnostic statuses

koorosh committed:
ui: Add tooltips for diagnostics statuses

- All code related to statements diagnostics are moved to
statements/diagnostics dir;
- Status badges are refactored to external component which is
responsible for displaying proper text of status and text for
tooltips.
- Tooltips have to be shown on some views and disabled on other.
To implement this, `enableTooltip` property is added. Particulary
it is needed to hide tooltip for "READY" status in Diagnostics tab
and show it on Statements List view.

Release note (admin ui): Show tooltips for statement diagnostic statuses

koorosh committed:
ui: Create HeaderSection shared component

It is common for pages to have header with 'back' navigation link.
HeaderSection component wraps page header with logic of navigation
back to specified page via props.

Release note: None

koorosh committed:
ui: Extend Statements table with Diagnostics column

Current changes add column which doesn't consume any
data and always shows 'Activate' link.
It is starting point for next changes:
- providing diagnostics statuses;
- displaying modal for requesting diagnostics;

Release note (admin ui): Add Diagnostics column to Statements Page

koorosh committed:
ui: Add Modal component

- Modal component is a wrapper around Ant.design Modal component
with customized styles;

Release note: None

koorosh committed:
ui: Integrate ActivateDiagnosticsModal with Statements page

ActivateDiagnosticsModal is implemented with total encapsulation
of modal's state and connected props to store into component.
- ActivateDiagnosticsModal component instance exposes `showModalFor`
method through forwardedRef that makes possible to manage open/closed
state of modal inside of component and parent component can trigger
where modal should be opened and provide required arguments.
Current pattern is more anti-pattern but with this particular case
it make easy to separate responsibilities between components.
- ActivateDiagnosticsModal component is connected to store and
dispatches action to request new diagnostics;

Statements table is based on SortedTable component which doesn't
provide easy way to handle clicks on rows (due to some magic
bootstrapping process). This required to propagate method which
triggers modal opening through several functions outside of
component.

Release note (admin ui change): Open Activate Diagnostics modal from Statements page (Activate link)

koorosh committed:
ui: Rename Link component to Anchor

Release note: None

koorosh committed:
ui: Integrate requestStatementDiagnostics API

Replace faked API with real implementation to
request statement diagnostics.

Release note: None

koorosh committed:
ui: Add selectDiagnosticRequestsByStatementFingerprint selector

Release note: None

koorosh committed:
ui: Connect Statement Diagnostics API

Before, Statements diagnostics view used faked API
to simulate responses from server.
Now redux API has changed to comply with actual
implementation of API.

Release note: None

koorosh committed:
ui: Connect Statement Diagnostics API

Before, Statements diagnostics view used faked API
to simulate responses from server.
Now redux API has changed to comply with actual
implementation of API.

Release note: None

koorosh committed:
ui: Connect Statement Diagnostics API wit diagnostics table component

Replacing faked API responses with real implementation required some
redesigns with following important changes:

- Requesting diagnosticsRequests initiated when DiagnosticsView is opened.
Then to request new diagnostics requests before existing data is invalidated,
we need to manually invalidate and then refresh data.

- Current backend API doesn't support Error state for requested data so only
"Ready", and "Waiting for query" statuses are possible.

Release note: None

koorosh committed:
ui: Add DownloadFile component

DownloadFile component accepts file name and content and
then initiates file download from client.
It works as a wrapper to catch onClick event on wrapped
component.

Release note: None

koorosh committed:
ui: Refactor Statement diagnostics action ()redux) names

Release note: None

koorosh committed:
ui: Add traces downloading on Statement Diagnostics tab

- Add ability to download json file with tracing data
- Refactor code to comply with changed backend API
- add `long` package because exposed interface for
getting statement diagnostics require diagnostics id
as Long type. Current library provides compatible
implementation of Long type with protobuf.js interfaces.

Release note: None

koorosh committed:
ui: Add Statement Diagnostics History page

- Page contains list of all available statement diagnostics reports
- Added link on Debug page to reference on Statement Diagnostics History page
- Added link on Statement Diagnostics tab to reference on Statement Diagnostics History page

Release note (admin ui change): Statement Diagnostics History page added

koorosh committed:
ui: Add links on Statement diagnostics documentation

Release note: None

koorosh committed:
ui: Remove View Trace button as functionality is not ready

Release note: None

koorosh committed:
ui: Apply sorting for Statement Diagnostics table and Diagnostics History page

- Sort by 'Activated On' column by default

Release note (admin ui change): Statement Diagnostics table, default sorting

koorosh committed:
ui: AutoClose AlertMessage on timeout

Before, alert messages could be closed only manually.
With current implementation, Alert component allows
to specify if it can be closed with timeout (6 seconds),
and if it can be closed manually (hide close button on alert).

Release note: None

koorosh committed:
ui: Show alert message when user requests new diagnostics report

- Show success alert message on Statements and Diagnostics View pages
when request is successful. Success message will be closed automatically
after 6 sec.
- In case of error - show failed message and requires to be closed manually.
- All messages will be closed when user leaves page.

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/45799)
<!-- Reviewable:end -->

koorosh committed:
ui: Fit Diagnostics column in Statements table

Statements table columns were styled to occupy all
width and adding one more column stretched table
width beyond the screen. To fit it, styles for columns
with bar charts were changed (decreased min width)

Release note: None

koorosh committed:
ui: Rename tracing file to statement-diagnostics.json

Release note: None

koorosh committed:
ui: Add handlers for downloading file from Statement Diagnostics History page

Logic for downloading file with traces was missed
on Diagnostics History page. It is the same functionality
as on Statement Diagnostics tab.

Release note: None

dhartunian commented:
All pre-reqs except backend change have been merged.

dhartunian commented on pkg/ui/src/views/statements/diagnostics/diagnosticStatuses.ts:
```diff
@@ -0,0 +1,11 @@
+// by the Apache License, Version 2.0, included in the file
+// licenses/APL.txt.
+
+export type DiagnosticStatuses = "READY" | "WAITING FOR QUERY" | "ERROR";
```
nit: this is duplicated in `diagnosticsUtils.ts`

dhartunian reviewed (approved):
Great work @koorosh! Thank you for including tests.

One minor nit for now although I didn't go through very carefully yet. I will revisit for now and we can revisit pieces as we expand this functionality.

dhartunian committed:
ui: small lint errors and redundant type

Release note: none
